vaadin-checkbox-flow/src/main/java/com/vaadin/flow/component/checkbox/CheckboxGroup.java

```diff
@@ -356,18 +362,21 @@ private void reset() {
         removeAll();
         clear();
 
-        final AtomicInteger itemsCount = new AtomicInteger(0);
-        getDataProvider().fetch(new Query<>()).map(this::createCheckBox)
-                .forEach(component -> {
-                    add(component);
-                    itemsCount.incrementAndGet();
-                });
-
-        final int newDataSize = itemsCount.get();
-        if (dataSize != newDataSize) {
-            dataSize = newDataSize;
-            fireEvent(new SizeChangeEvent<>(this, newDataSize));
+        synchronized (dataProvider) {
+            final AtomicInteger itemCounter = new AtomicInteger(0);
+            getDataProvider().fetch(new Query<>()).map(this::createCheckBox)
+                    .forEach(component -> {
+                        add(component);
+                        itemCounter.incrementAndGet();
+                    });
+            lastFetchedDataSize = itemCounter.get();
         }
+
+        runBeforeClientResponse(ui -> {
+            // Size event is fired before client response so as to avoid
+            // multiple size change events during server round trips
+            fireSizeEvent();
```

Comment:
```suggestion
        if(sizeRequest == null) {
            // Size event is fired before client response so as to avoid
            // multiple size change events during server round trips
            sizeRequest = context -> [
                fireSizeEvent();
                sizeRequest = null;
            }
            runBeforeClientResponse(ui ->sizeRequest());
        }
```